pkg/broker/ingress/multi_topic_decouple_sink.go

```diff
@@ -53,27 +63,81 @@ type multiTopicDecoupleSink struct {
-	return m.client.Send(ctx, event)
+
+	dt := extensions.FromSpanContext(trace.FromContext(ctx).SpanContext())
+	msg := new(pubsub.Message)
+	if err := cepubsub.WritePubSubMessage(ctx, binding.ToMessage(&event), msg, dt.WriteTransformer()); err != nil {
+		return err
+	}
+
+	_, err = topic.Publish(ctx, msg).Get(ctx)
+	return err
 }
 
 // getTopicForBroker finds the corresponding decouple topic for the broker from the mounted broker configmap volume.
-func (m *multiTopicDecoupleSink) getTopicForBroker(ns, broker string) (string, error) {
-	brokerConfig, ok := m.brokerConfig.GetBroker(ns, broker)
+func (m *multiTopicDecoupleSink) getTopicForBroker(broker types.NamespacedName) (*pubsub.Topic, error) {
+	topicID, err := m.getTopicIDForBroker(broker)
+	if err != nil {
+		return nil, err
+	}
+
+	if topic, ok := m.getExistingTopic(broker); ok {
+		// Check that the broker's topic ID hasn't changed.
+		if topic.ID() == topicID {
+			return topic, nil
+		}
+	}
+
+	// Topic needs to be created or updated.
+	return m.updateTopicForBroker(broker)
+}
+
+func (m *multiTopicDecoupleSink) updateTopicForBroker(broker types.NamespacedName) (*pubsub.Topic, error) {
+	m.topicsMut.Lock()
+	defer m.topicsMut.Unlock()
+	// Fetch latest decouple topic ID under lock.
+	topicID, err := m.getTopicIDForBroker(broker)
+	if err != nil {
+		return nil, err
+	}
+
+	if topic, ok := m.topics[broker]; ok {
+		if topic.ID() == topicID {
+			// Topic already updated.
+			return topic, nil
+		} else {
+			// Stop old topic.
+			m.topics[broker].Stop()
+		}
+	}
+	topic := m.pubsub.Topic(topicID)
+	m.topics[broker] = topic
+	return topic, nil
+}
+
+func (m *multiTopicDecoupleSink) getTopicIDForBroker(broker types.NamespacedName) (string, error) {
+	brokerConfig, ok := m.brokerConfig.GetBroker(broker.Namespace, broker.Name)
 	if !ok {
 		// There is an propagation delay between the controller reconciles the broker config and
 		// the config being pushed to the configmap volume in the ingress pod. So sometimes we return
 		// an error even if the request is valid.
-		m.logger.Warn("config is not found for", zap.Any("ns", ns), zap.Any("broker", broker))
-		return "", fmt.Errorf("%q/%q: %w", ns, broker, ErrNotFound)
+		m.logger.Warn("config is not found for", zap.String("broker", broker.String()))
+		return "", fmt.Errorf("%q: %w", broker, ErrNotFound)
 	}
 	if brokerConfig.DecoupleQueue == nil || brokerConfig.DecoupleQueue.Topic == "" {
 		m.logger.Error("DecoupleQueue or topic missing for broker, this should NOT happen.", zap.Any("brokerConfig", brokerConfig))
-		return "", fmt.Errorf("decouple queue of %q/%q: %w", ns, broker, ErrIncomplete)
+		return "", fmt.Errorf("decouple queue of %q: %w", broker, ErrIncomplete)
 	}
 	return brokerConfig.DecoupleQueue.Topic, nil
 }
+
+func (m *multiTopicDecoupleSink) getExistingTopic(broker types.NamespacedName) (*pubsub.Topic, bool) {
+	m.topicsMut.RLock()
+	defer m.topicsMut.RUnlock()
+	topic, ok := m.topics[broker]
```

Comment:
nit: `return m.topics[broker]`

Reply:
This does not work since this is a special case in go where the result is overloaded as both a 1 and 2 value expression and go will not perform type inference.